### ISSUE TITLE
cpuset controller can't be enabled if processed owned by different UID are in cgroup

### ISSUE THREAD

**Issue by MoritzS** [systemd/systemd#18293]
**systemd version the issue has been seen with**
 > systemd 247 (247.2-1-arch)
+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)
 > Linux myhostname 5.10.7-arch1-1 #1 SMP PREEMPT Wed, 13 Jan 2021 12:02:01 +0000 x86_64 GNU/Linux

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > When using Delegate=cpuset in user@.service.d/override.conf the cgroup for the user@.service should get the cpuset controller.

**Unexpected behaviour you saw**
 > Sometimes, i.e. not consistently, the cpuset controller is not set in /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control.

**Steps to reproduce the problem**

I am using `Delegate=` in `user@.service` to be able to use MemoryHigh, MemoryMax, AllowedCPUs, etc. in user units. I configured it as follows:

* Kernel command line has `systemd.unified_cgroup_hierarchy=1` to disable hybrid mode and enable v2
* `/etc/systemd/system/user@.service.d/override.conf` has these contents:
```
[Service]
Delegate=cpuset
Delegate=memory
Delegate=pids
```
* I configure a user slice (called "devenv") like this (note that this is in two separate commands, if that matters):
```
$ systemctl --user --runtime set-property devenv.slice \
            MemoryHigh=12G \
            MemoryMax=100%
$ systemctl --user --runtime set-property devenv.slice "AllowedCPUs=0-4 6-16 18-23"
```
* I start the user slice with `systemd-run --user --shell --slice=devenv.slice`

Most of the time, this works correctly, e.g., by checking `Cpus_allowed_list` in `/proc/self/status` from within the slice). However sometimes the cpuset is not being set. When this happens, I noticed that `/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control` contains only `memory pids` and not `cpuset`. When I manually add it by writing `+cpuset` to it as root and restart the slice, it works again.

**Comment by Werkov** [comment]
Thanks for the descriptive report, @MoritzS. The non-determinism is interesting, let me ask more then. When you reach the state when the `cpuset` controller is missing in `cgroup.subtree_control`, could you please capture the output of
```
systemctl show user@1000.service
systemctl dump
```
and share it. You don't need to share all dump, just `user@1000.service`.
Also, if you enable systemd debug logging (before `user@1000.service` starts (e.g. `systemd.log_level=debug` on kernel cmdline)), can you see any suspicious (cgroup related) messages in the log preceding failed `user@1000.service` delegation?

**Comment by MoritzS** [comment]
This is the output of `systemctl show user@1000.service` in the failure case
<details>
<summary>(Click to expand)</summary>
<p>

```
Type=notify
Restart=no
NotifyAccess=main
RestartUSec=100ms
TimeoutStartUSec=1min 30s
TimeoutStopUSec=2min
TimeoutAbortUSec=2min
TimeoutStartFailureMode=terminate
TimeoutStopFailureMode=terminate
RuntimeMaxUSec=infinity
WatchdogUSec=0
WatchdogTimestampMonotonic=0
RootDirectoryStartOnly=no
RemainAfterExit=no
GuessMainPID=yes
MainPID=900
ControlPID=0
FileDescriptorStoreMax=0
NFileDescriptorStore=0
StatusText=Startup finished in 120ms.
StatusErrno=0
Result=success
ReloadResult=success
CleanResult=success
UID=1000
GID=1000
NRestarts=0
OOMPolicy=continue
ExecMainStartTimestamp=Thu 2021-01-21 09:30:32 CET
ExecMainStartTimestampMonotonic=44979922
ExecMainExitTimestampMonotonic=0
ExecMainPID=900
ExecMainCode=0
ExecMainStatus=0
ExecStart={ path=/usr/lib/systemd/systemd ; argv[]=/usr/lib/systemd/systemd --user ; ignore_errors=no ; start_time=[Thu 2021-01-21 09:30:32 CET] ; stop_time=[n/a] ; pid=900 ; code=(null) ; status=0/0 }
ExecStartEx={ path=/usr/lib/systemd/systemd ; argv[]=/usr/lib/systemd/systemd --user ; flags= ; start_time=[Thu 2021-01-21 09:30:32 CET] ; stop_time=[n/a] ; pid=900 ; code=(null) ; status=0/0 }
Slice=user-1000.slice
ControlGroup=/user.slice/user-1000.slice/user@1000.service
MemoryCurrent=3947757568
CPUUsageNSec=239022051000
EffectiveCPUs=0-23
EffectiveMemoryNodes=0
TasksCurrent=776
IPIngressBytes=[no data]
IPIngressPackets=[no data]
IPEgressBytes=[no data]
IPEgressPackets=[no data]
IOReadBytes=18446744073709551615
IOReadOperations=18446744073709551615
IOWriteBytes=18446744073709551615
IOWriteOperations=18446744073709551615
Delegate=yes
DelegateControllers=cpuset memory pids
CPUAccounting=yes
CPUWeight=[not set]
StartupCPUWeight=[not set]
CPUShares=[not set]
StartupCPUShares=[not set]
CPUQuotaPerSecUSec=infinity
CPUQuotaPeriodUSec=infinity
AllowedCPUs=
AllowedMemoryNodes=
IOAccounting=no
IOWeight=[not set]
StartupIOWeight=[not set]
BlockIOAccounting=no
BlockIOWeight=[not set]
StartupBlockIOWeight=[not set]
MemoryAccounting=yes
DefaultMemoryLow=0
DefaultMemoryMin=0
MemoryMin=0
MemoryLow=0
MemoryHigh=infinity
MemoryMax=infinity
MemorySwapMax=infinity
MemoryLimit=infinity
DevicePolicy=auto
TasksAccounting=yes
TasksMax=infinity
IPAccounting=no
ManagedOOMSwap=auto
ManagedOOMMemoryPressure=auto
ManagedOOMMemoryPressureLimitPercent=0%
UMask=0022
LimitCPU=infinity
LimitCPUSoft=infinity
LimitFSIZE=infinity
LimitFSIZESoft=infinity
LimitDATA=infinity
LimitDATASoft=infinity
LimitSTACK=infinity
LimitSTACKSoft=8388608
LimitCORE=infinity
LimitCORESoft=infinity
LimitRSS=infinity
LimitRSSSoft=infinity
LimitNOFILE=524288
LimitNOFILESoft=1024
LimitAS=infinity
LimitASSoft=infinity
LimitNPROC=256838
LimitNPROCSoft=256838
LimitMEMLOCK=65536
LimitMEMLOCKSoft=65536
LimitLOCKS=infinity
LimitLOCKSSoft=infinity
LimitSIGPENDING=256838
LimitSIGPENDINGSoft=256838
LimitMSGQUEUE=819200
LimitMSGQUEUESoft=819200
LimitNICE=0
LimitNICESoft=0
LimitRTPRIO=0
LimitRTPRIOSoft=0
LimitRTTIME=infinity
LimitRTTIMESoft=infinity
RootHashSignature=
OOMScoreAdjust=0
CoredumpFilter=0x33
Nice=0
IOSchedulingClass=0
IOSchedulingPriority=0
CPUSchedulingPolicy=0
CPUSchedulingPriority=0
CPUAffinity=
CPUAffinityFromNUMA=no
NUMAPolicy=n/a
NUMAMask=
TimerSlackNSec=50000
CPUSchedulingResetOnFork=no
NonBlocking=no
StandardInput=null
StandardInputData=
StandardOutput=journal
StandardError=inherit
TTYReset=no
TTYVHangup=no
TTYVTDisallocate=no
SyslogPriority=30
SyslogLevelPrefix=yes
SyslogLevel=6
SyslogFacility=3
LogLevelMax=-1
LogRateLimitIntervalUSec=0
LogRateLimitBurst=0
SecureBits=0
CapabilityBoundingSet=cap_chown cap_dac_override cap_dac_read_search cap_fowner cap_fsetid cap_kill cap_setgid cap_setuid cap_setpcap cap_linux_immutable cap_net_bind_service cap_net_broadcast cap_net_admin cap_net_raw cap_ipc_lock cap_ipc_owner cap_sys_module cap_sys_rawio cap_sys_chroot cap_sys_ptrace cap_sys_pacct cap_sys_admin cap_sys_boot cap_sys_nice cap_sys_resource cap_sys_time cap_sys_tty_config cap_mknod cap_lease cap_audit_write cap_audit_control cap_setfcap cap_mac_override cap_mac_admin cap_syslog cap_wake_alarm cap_block_suspend cap_audit_read cap_perfmon cap_bpf cap_checkpoint_restore
AmbientCapabilities=
User=1000
DynamicUser=no
RemoveIPC=no
PAMName=systemd-user
MountFlags=
PrivateTmp=no
PrivateDevices=no
ProtectClock=no
ProtectKernelTunables=no
ProtectKernelModules=no
ProtectKernelLogs=no
ProtectControlGroups=no
PrivateNetwork=no
PrivateUsers=no
PrivateMounts=no
ProtectHome=no
ProtectSystem=no
SameProcessGroup=no
UtmpMode=init
IgnoreSIGPIPE=yes
NoNewPrivileges=no
SystemCallErrorNumber=2147483646
LockPersonality=no
RuntimeDirectoryPreserve=no
RuntimeDirectoryMode=0755
StateDirectoryMode=0755
CacheDirectoryMode=0755
LogsDirectoryMode=0755
ConfigurationDirectoryMode=0755
TimeoutCleanUSec=infinity
MemoryDenyWriteExecute=no
RestrictRealtime=no
RestrictSUIDSGID=no
RestrictNamespaces=no
MountAPIVFS=no
KeyringMode=inherit
ProtectProc=default
ProcSubset=all
ProtectHostname=no
KillMode=mixed
KillSignal=15
RestartKillSignal=15
FinalKillSignal=9
SendSIGKILL=yes
SendSIGHUP=no
WatchdogSignal=6
Id=user@1000.service
Names=user@1000.service
Requires=user-runtime-dir@1000.service user-1000.slice sysinit.target
WantedBy=session-2.scope
Conflicts=shutdown.target
Before=shutdown.target session-2.scope
After=systemd-journald.socket dbus.service user-1000.slice sysinit.target systemd-user-sessions.service basic.target user-runtime-dir@1000.service
Documentation="man:user@.service(5)"
Description=User Manager for UID 1000
LoadState=loaded
ActiveState=active
FreezerState=running
SubState=running
FragmentPath=/usr/lib/systemd/system/user@.service
DropInPaths=/etc/systemd/system/user@.service.d/override.conf
UnitFileState=static
UnitFilePreset=disabled
StateChangeTimestamp=Thu 2021-01-21 09:30:33 CET
StateChangeTimestampMonotonic=45121922
InactiveExitTimestamp=Thu 2021-01-21 09:30:32 CET
InactiveExitTimestampMonotonic=44980087
ActiveEnterTimestamp=Thu 2021-01-21 09:30:33 CET
ActiveEnterTimestampMonotonic=45121922
ActiveExitTimestampMonotonic=0
InactiveEnterTimestampMonotonic=0
CanStart=yes
CanStop=yes
CanReload=no
CanIsolate=no
CanFreeze=yes
StopWhenUnneeded=no
RefuseManualStart=no
RefuseManualStop=no
AllowIsolate=no
DefaultDependencies=yes
OnFailureJobMode=replace
IgnoreOnIsolate=yes
NeedDaemonReload=no
JobTimeoutUSec=infinity
JobRunningTimeoutUSec=infinity
JobTimeoutAction=none
ConditionResult=yes
AssertResult=yes
ConditionTimestamp=Thu 2021-01-21 09:30:32 CET
ConditionTimestampMonotonic=44979248
AssertTimestamp=Thu 2021-01-21 09:30:32 CET
AssertTimestampMonotonic=44979250
Transient=no
Perpetual=no
StartLimitIntervalUSec=10s
StartLimitBurst=5
StartLimitAction=none
FailureAction=none
SuccessAction=none
InvocationID=f99a694d37614ab1ae083aabe19fff97
CollectMode=inactive
```
</p>
</details>

And this is the diff to another output of `systemctl show user@1000.service` where it works (--- is the version from above where it failed, +++ is the dump of when it works).

<details>
<summary>(Click to expand)</summary>
<p>

```diff
--- user@1000.service	2021-01-21 09:57:52.976175896 +0100
+++ /proc/self/fd/11	2021-01-22 18:01:15.184606670 +0100
@@ -13,11 +13,11 @@
 RootDirectoryStartOnly=no
 RemainAfterExit=no
 GuessMainPID=yes
-MainPID=900
+MainPID=891
 ControlPID=0
 FileDescriptorStoreMax=0
 NFileDescriptorStore=0
-StatusText=Startup finished in 120ms.
+StatusText=Startup finished in 124ms.
 StatusErrno=0
 Result=success
 ReloadResult=success
@@ -26,21 +26,21 @@
 GID=1000
 NRestarts=0
 OOMPolicy=continue
-ExecMainStartTimestamp=Thu 2021-01-21 09:30:32 CET
-ExecMainStartTimestampMonotonic=44979922
+ExecMainStartTimestamp=Fri 2021-01-22 17:54:25 CET
+ExecMainStartTimestampMonotonic=10753018
 ExecMainExitTimestampMonotonic=0
-ExecMainPID=900
+ExecMainPID=891
 ExecMainCode=0
 ExecMainStatus=0
-ExecStart={ path=/usr/lib/systemd/systemd ; argv[]=/usr/lib/systemd/systemd --user ; ignore_errors=no ; start_time=[Thu 2021-01-21 09:30:32 CET] ; stop_time=[n/a] ; pid=900 ; code=(null) ; status=0/0 }
-ExecStartEx={ path=/usr/lib/systemd/systemd ; argv[]=/usr/lib/systemd/systemd --user ; flags= ; start_time=[Thu 2021-01-21 09:30:32 CET] ; stop_time=[n/a] ; pid=900 ; code=(null) ; status=0/0 }
+ExecStart={ path=/usr/lib/systemd/systemd ; argv[]=/usr/lib/systemd/systemd --user ; ignore_errors=no ; start_time=[Fri 2021-01-22 17:54:25 CET] ; stop_time=[n/a] ; pid=891 ; code=(null) ; status=0/0 }
+ExecStartEx={ path=/usr/lib/systemd/systemd ; argv[]=/usr/lib/systemd/systemd --user ; flags= ; start_time=[Fri 2021-01-22 17:54:25 CET] ; stop_time=[n/a] ; pid=891 ; code=(null) ; status=0/0 }
 Slice=user-1000.slice
 ControlGroup=/user.slice/user-1000.slice/user@1000.service
-MemoryCurrent=3947757568
-CPUUsageNSec=239022051000
+MemoryCurrent=1677090816
+CPUUsageNSec=83935362000
 EffectiveCPUs=0-23
 EffectiveMemoryNodes=0
-TasksCurrent=776
+TasksCurrent=390
 IPIngressBytes=[no data]
 IPIngressPackets=[no data]
 IPEgressBytes=[no data]
@@ -199,7 +199,7 @@
 WantedBy=session-2.scope
 Conflicts=shutdown.target
 Before=shutdown.target session-2.scope
-After=systemd-journald.socket dbus.service user-1000.slice sysinit.target systemd-user-sessions.service basic.target user-runtime-dir@1000.service
+After=systemd-journald.socket systemd-user-sessions.service dbus.service user-runtime-dir@1000.service user-1000.slice basic.target sysinit.target
 Documentation="man:user@.service(5)"
 Description=User Manager for UID 1000
 LoadState=loaded
@@ -210,12 +210,12 @@
 DropInPaths=/etc/systemd/system/user@.service.d/override.conf
 UnitFileState=static
 UnitFilePreset=disabled
-StateChangeTimestamp=Thu 2021-01-21 09:30:33 CET
-StateChangeTimestampMonotonic=45121922
-InactiveExitTimestamp=Thu 2021-01-21 09:30:32 CET
-InactiveExitTimestampMonotonic=44980087
-ActiveEnterTimestamp=Thu 2021-01-21 09:30:33 CET
-ActiveEnterTimestampMonotonic=45121922
+StateChangeTimestamp=Fri 2021-01-22 17:54:25 CET
+StateChangeTimestampMonotonic=10898201
+InactiveExitTimestamp=Fri 2021-01-22 17:54:25 CET
+InactiveExitTimestampMonotonic=10753188
+ActiveEnterTimestamp=Fri 2021-01-22 17:54:25 CET
+ActiveEnterTimestampMonotonic=10898201
 ActiveExitTimestampMonotonic=0
 InactiveEnterTimestampMonotonic=0
 CanStart=yes
@@ -236,10 +236,10 @@
 JobTimeoutAction=none
 ConditionResult=yes
 AssertResult=yes
-ConditionTimestamp=Thu 2021-01-21 09:30:32 CET
-ConditionTimestampMonotonic=44979248
-AssertTimestamp=Thu 2021-01-21 09:30:32 CET
-AssertTimestampMonotonic=44979250
+ConditionTimestamp=Fri 2021-01-22 17:54:25 CET
+ConditionTimestampMonotonic=10752345
+AssertTimestamp=Fri 2021-01-22 17:54:25 CET
+AssertTimestampMonotonic=10752347
 Transient=no
 Perpetual=no
 StartLimitIntervalUSec=10s
@@ -247,5 +247,5 @@
 StartLimitAction=none
 FailureAction=none
 SuccessAction=none
-InvocationID=f99a694d37614ab1ae083aabe19fff97
+InvocationID=ecae9fb8721749cbaf3465402961c94c
 CollectMode=inactive
```

</p>
</details>

The only significant difference is the order of the units in `After=`. Is that relevant?

I will try to get the debug log from when it fails.

**Comment by Werkov** [comment]
Nothing wrong here (`DelegetaControllers=` is set as expected). Do you also have the `systemctl dump` data (there are some more details about `user@1000.service`)?

**Comment by MoritzS** [comment]
Here is the output of `systemd-analyze dump` for `user@1000.service`:

<details>
<summary>(Click to expand)</summary>
<p>

```
-> Unit user@1000.service:
	Description: User Manager for UID 1000
	Instance: 1000
	Unit Load State: loaded
	Unit Active State: active
	State Change Timestamp: Tue 2021-01-26 09:25:57 CET
	Inactive Exit Timestamp: Tue 2021-01-26 09:25:57 CET
	Active Enter Timestamp: Tue 2021-01-26 09:25:57 CET
	Active Exit Timestamp: n/a
	Inactive Enter Timestamp: n/a
	May GC: no
	Need Daemon Reload: no
	Transient: no
	Perpetual: no
	Garbage Collection Mode: inactive
	Slice: user-1000.slice
	CGroup: /user.slice/user-1000.slice/user@1000.service
	CGroup realized: yes
	CGroup realized mask: cpuset memory pids
	CGroup own mask: cpuset memory pids
	CGroup delegate mask: cpuset memory pids
	Invocation ID: 6674f0e625174bc8965c899bba4fe76f
	Documentation: man:user@.service(5)
	Fragment Path: /usr/lib/systemd/system/user@.service
	DropIn Path: /etc/systemd/system/user@.service.d/override.conf
	Condition Timestamp: Tue 2021-01-26 09:25:57 CET
	Condition Result: yes
	Assert Timestamp: Tue 2021-01-26 09:25:57 CET
	Assert Result: yes
	Requires: user-1000.slice (origin-file)
	Requires: user-runtime-dir@1000.service (origin-file)
	Requires: sysinit.target (origin-default)
	WantedBy: session-2.scope (destination-file)
	Conflicts: shutdown.target (origin-default)
	Before: shutdown.target (origin-default)
	Before: session-2.scope (destination-file)
	After: systemd-journald.socket (origin-file)
	After: sysinit.target (origin-default)
	After: user-runtime-dir@1000.service (origin-file)
	After: dbus.service (origin-file)
	After: systemd-user-sessions.service (origin-file)
	After: basic.target (origin-default)
	After: user-1000.slice (origin-file)
	References: user-1000.slice (origin-file)
	References: shutdown.target (origin-default)
	References: systemd-journald.socket (origin-file)
	References: basic.target (origin-default)
	References: sysinit.target (origin-default)
	References: user-runtime-dir@1000.service (origin-file)
	References: systemd-user-sessions.service (origin-file)
	References: dbus.service (origin-file)
	ReferencedBy: session-2.scope (destination-file)
	StopWhenUnneeded: no
	RefuseManualStart: no
	RefuseManualStop: no
	DefaultDependencies: yes
	OnFailureJobMode: replace
	IgnoreOnIsolate: yes
	Service State: running
	Result: success
	Reload Result: success
	Clean Result: success
	PermissionsStartOnly: no
	RootDirectoryStartOnly: no
	RemainAfterExit: no
	GuessMainPID: yes
	Type: notify
	Restart: no
	NotifyAccess: main
	NotifyState: ready
	OOMPolicy: continue
	Main PID: 3277
	Main PID Known: yes
	Main PID Alien: no
	RestartSec: 100ms
	TimeoutStartSec: 1min 30s
	TimeoutStopSec: 2min
	TimeoutStartFailureMode: terminate
	TimeoutStopFailureMode: terminate
	RuntimeMaxSec: infinity
	WatchdogSec: 0
	KillMode: mixed
	KillSignal: SIGTERM
	RestartKillSignal: SIGTERM
	FinalKillSignal: SIGKILL
	SendSIGKILL: yes
	SendSIGHUP: no
	UMask: 0022
	WorkingDirectory: /
	RootDirectory: /
	NonBlocking: no
	PrivateTmp: no
	PrivateDevices: no
	ProtectKernelTunables: no
	ProtectKernelModules: no
	ProtectKernelLogs: no
	ProtectClock: no
	ProtectControlGroups: no
	PrivateNetwork: no
	PrivateUsers: no
	ProtectHome: no
	ProtectSystem: no
	MountAPIVFS: no
	IgnoreSIGPIPE: yes
	MemoryDenyWriteExecute: no
	RestrictRealtime: no
	RestrictSUIDSGID: no
	KeyringMode: inherit
	ProtectHostname: no
	ProtectProc: default
	ProcSubset: all
	RuntimeDirectoryPreserve: no
	RuntimeDirectoryMode: 0755
	StateDirectoryMode: 0755
	CacheDirectoryMode: 0755
	LogsDirectoryMode: 0755
	ConfigurationDirectoryMode: 0755
	TimeoutCleanSec: infinity
	LimitNOFILE: 524288
	LimitNOFILESoft: 1024
	LimitMEMLOCK: 65536
	LimitMEMLOCKSoft: 65536
	StandardInput: null
	StandardOutput: journal
	StandardError: inherit
	SyslogFacility: daemon
	SyslogLevel: info
	User: 1000
	DynamicUser: no
	PAMName: systemd-user
	LockPersonality: no
	SystemCallErrorNumber: kill
	-> ExecStart:
		Command Line: /usr/lib/systemd/systemd --user
			PID: 3277
			Start Timestamp: Tue 2021-01-26 09:25:57 CET
	Status Text: Startup finished in 95ms.
	CPUAccounting: yes
	IOAccounting: no
	BlockIOAccounting: no
	MemoryAccounting: yes
	TasksAccounting: yes
	IPAccounting: no
	CPUWeight: 18446744073709551615
	StartupCPUWeight: 18446744073709551615
	CPUShares: 18446744073709551615
	StartupCPUShares: 18446744073709551615
	CPUQuotaPerSecSec: infinity
	CPUQuotaPeriodSec: infinity
	AllowedCPUs: 
	AllowedMemoryNodes: 
	IOWeight: 18446744073709551615
	StartupIOWeight: 18446744073709551615
	BlockIOWeight: 18446744073709551615
	StartupBlockIOWeight: 18446744073709551615
	DefaultMemoryMin: 0
	DefaultMemoryLow: 0
	MemoryMin: 0
	MemoryLow: 0
	MemoryHigh: 18446744073709551615
	MemoryMax: 18446744073709551615
	MemorySwapMax: 18446744073709551615
	MemoryLimit: 18446744073709551615
	TasksMax: 18446744073709551615
	DevicePolicy: auto
	DisableControllers: 
	Delegate: yes
	ManagedOOMSwap: auto
	ManagedOOMMemoryPressure: auto
	ManagedOOMMemoryPressureLimitPercent: 0%
	DelegateControllers: cpuset memory pids
```

</p>
</details>

And this is the output of `systemd-analyze --user dump` for `devenv.slice`:

<details>
<summary>(Click to expand)</summary>
<p>

```
-> Unit devenv.slice:
	Description: devenv.slice
	Instance: n/a
	Unit Load State: loaded
	Unit Active State: active
	State Change Timestamp: Tue 2021-01-26 09:29:19 CET
	Inactive Exit Timestamp: Tue 2021-01-26 09:29:19 CET
	Active Enter Timestamp: Tue 2021-01-26 09:29:19 CET
	Active Exit Timestamp: n/a
	Inactive Enter Timestamp: n/a
	May GC: no
	Need Daemon Reload: yes
	Transient: no
	Perpetual: no
	Garbage Collection Mode: inactive
	Slice: -.slice
	CGroup: /user.slice/user-1000.slice/user@1000.service/devenv.slice
	CGroup realized: yes
	CGroup realized mask: cpuset memory pids
	CGroup enabled mask: memory pids
	CGroup own mask: cpuset memory pids
	CGroup members mask: memory pids
	Invocation ID: fbedd864f3014a3fb91197fed103e8fc
	DropIn Path: /run/user/1000/systemd/user.control/devenv.slice.d/50-AllowedCPUs.conf
	DropIn Path: /run/user/1000/systemd/user.control/devenv.slice.d/50-MemoryHigh.conf
	DropIn Path: /run/user/1000/systemd/user.control/devenv.slice.d/50-MemoryMaxScale.conf
	Condition Timestamp: Tue 2021-01-26 09:29:19 CET
	Condition Result: yes
	Assert Timestamp: Tue 2021-01-26 09:29:19 CET
	Assert Result: yes
	Requires: -.slice (origin-implicit)
	RequiredBy: run-u58.service (destination-file)
	Conflicts: shutdown.target (origin-default)
	Before: run-u58.service (destination-file)
	Before: shutdown.target (origin-default)
	After: -.slice (origin-implicit)
	References: -.slice (origin-implicit)
	References: shutdown.target (origin-default)
	ReferencedBy: run-u58.service (destination-file)
	StopWhenUnneeded: no
	RefuseManualStart: no
	RefuseManualStop: no
	DefaultDependencies: yes
	OnFailureJobMode: replace
	IgnoreOnIsolate: yes
	Slice State: active
	CPUAccounting: yes
	IOAccounting: no
	BlockIOAccounting: no
	MemoryAccounting: yes
	TasksAccounting: yes
	IPAccounting: no
	CPUWeight: 18446744073709551615
	StartupCPUWeight: 18446744073709551615
	CPUShares: 18446744073709551615
	StartupCPUShares: 18446744073709551615
	CPUQuotaPerSecSec: infinity
	CPUQuotaPeriodSec: infinity
	AllowedCPUs: 0-8 10-20 22-23
	AllowedMemoryNodes: 
	IOWeight: 18446744073709551615
	StartupIOWeight: 18446744073709551615
	BlockIOWeight: 18446744073709551615
	StartupBlockIOWeight: 18446744073709551615
	DefaultMemoryMin: 0
	DefaultMemoryLow: 0
	MemoryMin: 0
	MemoryLow: 0
	MemoryHigh: 12884901888
	MemoryMax: 67420635136
	MemorySwapMax: 18446744073709551615
	MemoryLimit: 18446744073709551615
	TasksMax: 18446744073709551615
	DevicePolicy: auto
	DisableControllers: 
	Delegate: no
	ManagedOOMSwap: auto
	ManagedOOMMemoryPressure: auto
	ManagedOOMMemoryPressureLimitPercent: 0%
```

</p>
</details>

There `CGroup enabled mask` and `CGroup members mask` seems to not include `cpuset`?

Let me know if you need any more parts of the dump.

**Comment by Werkov** [comment]
I can't see where the loophole could be so I was trying to reproduce your problem but to no avail. The `subtree_control ` always lists `cpuset` controller after I start a unit under `devenv.slice`.

My non-reproducer when I deliberately interleave delegation and (--user) slice start:
```
# cpuset is not delegated
# systemctl start user@1000.service # implicit via login
user: systemctl --user --runtime set-property devenv.slice "AllowedCPUs=..."
root: # edit override.conf, add cpuset controller
root: systemct daemon-reload
user: systemd-run --user --shell --slice=devenv.slice
cat /sys/fs/cgroup/user.slice/user-1000.slice/user\@1000.service/cgroup.subtree_control
```

@MoritzS Could you find a minimal set of steps leading to your issue (or a loop that yields that eventually)? And one more question, does `systemctl --user daemon-reload` fix the missing cpuset controller? TY

**Comment by MoritzS** [comment]
I tried rebooting multiple times, but it doesn't happen consistently unfortunately. `systemctl --user daemon-reload` also doesn't fix it.

However, I have been running with `systemd.log_level=debug` and got this:
```
systemd[921]: Failed to enable controller cpuset for /user.slice/user-1000.slice/user@1000.service (/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control): Operation not permitted
```

Here is the log output that was generated when I created the devenv slice: [log.txt](https://github.com/systemd/systemd/files/6206525/log.txt). Let me know if you need more log output. They only other mention of `cpuset` I found in the log is this:
```
systemd[1]: Controller 'cpuset' supported: yes
```

**Comment by Werkov** [comment]
> systemd[921]: Failed to enable controller cpuset for /user.slice/user-1000.slice/user@1000.service (/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control): Operation not permitted

This is very interesting. It implies that `open(2)(cgroup.subtree_control)` succeeded (otherwise there'd be an earlier debug message) but `write(2)` into the file apparently failed with -EPERM. So it's not wrong permissions on the file and I don't see a way this would originate from cgroup core code itself.

@MoritzS Any chance you have some SELinux, AppArmor or seccomp restrictive rules in place that'd prevent writing into `cgroup.subtree_control` from the user instance?

**Comment by MoritzS** [comment]
I don't have either of those security modules enabled. My devenv failed again today and I checked the permissions before I started the slice and got this:
```
$ ls -l /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control
-rw-r--r-- 1 moritz moritz 0 30. Apr 08:59 /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control
```
The permissions seem to be correct? Can the EPERM be caused by anything else? If I try to write `+cpuset` into that file manually as the user, I also get the EPERM. It only works as root. dmesg doesn't show me any kernel messages.

**Comment by Werkov** [comment]
Thanks for checking.
So what about this -- are there any tasks in the `user@1000.service` cgroup subtree that don't run as uid 1000? (That could fail the cpuset enabling attempt.)

**Comment by MoritzS** [comment]
Wow that actually appeared to be the problem. I looked at all processes in `user@1000.service` and found only one that was not running under my user.

For reference, this was the command I used:
```
systemctl status user@1000.service | awk 'match($0, /─ *[0-9]+/) { match($0, /[0-9]+/); print "ls -ld /proc/" substr($0, RSTART, RLENGTH) }' | bash
```

The process I found was running with uid root and comes from `xdg-document-portal.service` which is installed on my system as dependency to flatpak. So, when I started my devenv slice only after starting a flatpak app, it failed.

Now, is this a bug in flatpak or systemd?

**Comment by MoritzS** [comment]
Maybe an important note to add: After I manually stopped the `xdg-document-portal.service` and started my devenv slice, everything worked as expected.

**Comment by serverwentdown** [comment]
I experienced the exact same bug as mentioned.

**Comment by Werkov** [comment]
I don't think this is bug in flatpak or systemd per se. It's rather an unclear migration model of foreign tasks on the unified hierarchy. I've raised it in [cgroup ML](https://lore.kernel.org/r/20220623124944.2753-1-mkoutny@suse.com/).

**Comment by Steiniche** [comment]
Seeing the exact same bug in regards to running Podman which is also tracked over at Redhat https://bugzilla.redhat.com/show_bug.cgi?id=1897579 where the consensus seems to be that this is an systemd issue hence I am posting here.

Podman have a systemd restart service for restarting containers on reboots but if the `pids` cgroup controller is not available this will fail.

Bug is persistent in:
`systemd 239 (239-58.el8)
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=legacy`

In 239-58.el8 the bug could be fixed by adding  `/etc/systemd/system/user-.slice.d/override.conf` with the following content
```
[Slice]
Slice=user.slice

CPUAccounting=yes
MemoryAccounting=yes
IOAccounting=yes
TasksAccounting=yes
```
and running a `sudo systemctl daemon-reload`.

Almost always the pids and memory controllers would show up in `cat /sys/fs/cgroup/user.slice/user-$(id -u).slice/user@$(id -u).service/cgroup.controllers`

Bug seems to be worse in:
`systemd 239 (239-58.el8_6.4)
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=legacy`

In 239-58.el8_6.4 however it seems like the bug is more persistent and cannot be workaround with `sudo systemctl daemon-reload`.
If the pids controller does not show up the only way to fix it seems to be rebooting the machine until it works.

Have tried to also add a delegate.conf as suggested in the Redhat thread:
/etc/systemd/system/user@.service.d/delegate.conf
[Service]
Delegate=memory pids 

But this has no effect either.

I have also looked into the problem @MoritzS was experiencing but all procs are owned by 1000:1000

If any more debug information is needed to figure out why this happens please let me know and I will provide it.

**Comment by Werkov** [comment]
@Steiniche Unfortunately, the upstream tracker is usually used for two most recently released major systemd versions.

```
git log --no-merges --oneline v239..v251 src/core/cgroup.c src/shared/cgroup-setup.c | wc -l
221
```

(I know it's a distro with backports, so I'd check the state of these [commits](https://github.com/systemd/systemd/pull/15662/commits) and possibly [commit](github.com/systemd/systemd/commit/e00068e71fcf33ccd9190cb18d1539cd921cf444).)

> I have also looked into the problem @MoritzS was experiencing but all procs are > owned by 1000:1000

Thanks for checking, thus I'd qualify your issue as a different bug (feel free to file a new issue if you're convinced upstream versions reproduce the behavior).

**Comment by kaedwen** [comment]
This helped me out https://rootlesscontaine.rs/getting-started/common/cgroup2/

```
$ cat <<EOF | sudo tee /etc/systemd/system/user@.service.d/delegate.conf
[Service]
Delegate=cpu cpuset io memory pids
EOF
$ sudo systemctl daemon-reload
```

**Comment by Steiniche** [comment]
@Werkov thanks for clarifying that.

I agree that it is likely a different bug.

@kaedwen thanks for the sugegstion and most useful link.
However, it does not fix the problem as sometimes the controllers are simply not available after a reboot.

For others I have just tested with systemd systemd `250 (250-6.el9_0)`
`+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified`

Both with and without `delegate.conf` and `override.conf`

The issue seems to be completly fixed and the workarounds is not needed any more, if you only need `pids` and `memory`.

**Comment by escapety** [comment]
It seems that  systemd-250 have solved this problem, but it's hard for me to update the version of systemd and I want to know which commits fixed this problem then I can backport them to systemd-239 @Steiniche @Werkov , thanks.

**Comment by Steiniche** [comment]
For other people looking into this and stumbling upon this issue:

Check if SELinux is blocking access to the override.conf and delegate.conf.
This was the issue we were facing.

This can be checked by looking through `journalctl`

If you are seeing error messeges the fix is to run `restorecon -R /etc/systemd/system/` which will ensure that the owner etc. is correct for the file.

**Comment by poettering** [comment]
Werkov, what#s the state of this? did this get resolved in the kernel upstream? did your patch (or something related) get merged that addresses this? 

Can we close this here?

**Comment by arsham** [comment]
I just figured out a way to have the portal running and be able to use the delegation. I have noticed that starting the `xdg-document-portal.service`  while a unit it running using `AllowedCPUs`, for every attempt after that things are working as expected.

## Not Working
This is the situation that I couldn't set the `AllowedCPUs`:
```bash
systemctl --user start xdg-document-portal.service
systemd-run --user -p AllowedCPUs=2 -- my_command
```
# This Works

```bash
systemd-run --user -p AllowedCPUs=2 -- sleep 3
systemctl --user start xdg-document-portal.service
systemd-run --user -p AllowedCPUs=2 -- my_command
```

**Comment by Werkov** [comment]
Re: @arsham's workaround
When you enable cpuset controller first and the privileged process migrates into destination cgroup itself (or by action of another privileged process like PID 1) second, you don't need to enable/disable the controller with a privileged process in place, thus it works :+1: .
(But generally, let's stick to cpuset staying opt-in.)

> what#s the state of this? did this get resolved in the kernel upstream? did your patch (or something related) get merged that addresses this?
> 
> Can we close this here?

After one [last push](https://lore.kernel.org/r/20230703172741.25392-1-mkoutny@suse.com/), the kernel fix is scheduled for an upcoming version.

Thus nothing more to do here and closing. Thank all involved.
